### PR TITLE
Logcollector integration tests T0: Adapt logcollector integration tests for macOS

### LINKED ISSUE #1310: Adapt current logcollector integration tests for macOS
#### Issue information
| Epic 	| QA base branch 	| Wazuh branch 	|
|-	|-	|-	|
| https://github.com/wazuh/wazuh-qa/issues/1200 	| master 	|  4572-dev-uls-logs-enhancement |

## Issue definition

We need to adapt our new logcollector tests to be launched in macOS agents.

The configuration block for macOS would be:

```
<localfile>
  <log_format>oslog</log_format>
  <location>oslog</location>
  <query type="trace" level="debug">processImagePath CONTAINS[c] "com.apple.geod"</query>
</localfile>
```

### TODO
- [x] Review status of current test and compatibility with macOS. Make a list of test that doesn't work on macOS.
- [x] Adapt the failed tests to run in macOS systems.
- [x] Adapt `only-future-event`  to macos ([Check this issue](https://github.com/wazuh/wazuh/issues/8138))
- [x] Test_log_format:  add `macos` option
- [x] Test _location: add  `macos option`
- [x] Add `query` label to localfile on macos test cases when required.

### PR DESCRIPTION
**Pull request by Rebits**
|Related issue|
|---|
|Closes #1310 |

## Description

This PR adapt current Tier0 logcollector tests to MacOS systems. Also, some common logcollector fixtures and tests have been refactored so further testing should be required.

### Manager - CentOS

manager_logcollector

### Windows

query_windows

#### Log format 

log_format_macos

#### Location

macos_location

#### Query

query

## Documentation

No new tests have been added, so, all new changes on documentation are relative to new utils for refactoring or macOS agent testing.

logcollector_docu

utils_docu

### CODE REVIEW

File: deps/wazuh_testing/wazuh_testing/tools/utils.py [hunk 1]
```diff
@@ -102,10 +102,41 @@ def replace_in_file(filename, to_replace, replacement):
...
\ No newline at end of file
+    return ''.join(SystemRandom().choice(character_set) for _ in range(string_length))
+
+
+def lower_case_key_dictionary_array(array_dict):
+    """Given an array of dictionaries, create a copy of it with the keys of each dictionary in lower case.
+
+    Args:
+        array_dict (List): List of dictionaries.
+    """
```
Comment (BraulioV) [inline deps/wazuh_testing/wazuh_testing/tools/utils.py:133]: Returns field is missing.

File: deps/wazuh_testing/wazuh_testing/tools/utils.py [hunk 2]
```diff
@@ -102,10 +102,41 @@ def replace_in_file(filename, to_replace, replacement):
...
+    lower_case_array_dict = []
+
+    for element in array_dict:
+        lower_case_element = {}
+        for key, value in element.items():
+            lower_case_element[str(key).lower()] = value
+        lower_case_array_dict.append(lower_case_element)
+
+    return lower_case_array_dict
```
Comment (BraulioV) [inline deps/wazuh_testing/wazuh_testing/tools/utils.py:142]:
```suggestion
    return [{str(key).lower(): value for key, value in element.items() for element in array_dict}]
```

File: tests/integration/test_logcollector/conftest.py [hunk 3]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (BraulioV) [inline tests/integration/test_logcollector/conftest.py:1]: Remove this line

File: tests/integration/test_logcollector/test_configuration/test_basic_configuration_query.py [hunk 4]
```diff
@@ -64,6 +101,14 @@ def test_configuration_query_valid(get_configuration, configure_environment, res
...
     """
 
-    log_callback = logcollector.callback_eventchannel_analyzing('Security')
-    wazuh_log_monitor.start(timeout=5, callback=log_callback,
-                            error_message=logcollector.GENERIC_CALLBACK_ERROR_ANALYZING_EVENTCHANNEL)
\ No newline at end of file
+    configuration = get_configuration['metadata']
+    log_format = configuration['log_format']
+
+    if log_format == 'oslog':
```
Comment (BraulioV) [inline tests/integration/test_logcollector/test_configuration/test_basic_configuration_query.py:107]:
```suggestion
    if log_format == 'macos':
```

File: tests/integration/test_logcollector/test_configuration/test_basic_configuration_query.py [hunk 5]
```diff
@@ -5,48 +5,85 @@
...
+wazuh_configuration = 'wazuh_basic_configuration_query_macos.yaml'
+
+if sys.platform != 'win32' and sys.platform != 'darwin':
     pytestmark = [pytest.mark.skip, pytest.mark.tier(level=0)]
 else:
-    pytestmark = pytest.mark.tier(level=0)
+    pytestmark = [pytest.mark.tier(level=0)]
+    if sys.platform == 'darwin':
+        clauses = ['eventMessage', 'processImagePath', 'senderImagePath', 'subsystem', 'category']
+        location = log_format = 'oslog'
```
Comment (BraulioV) [inline tests/integration/test_logcollector/test_configuration/test_basic_configuration_query.py:26]: This should be `macos`. Also, check this line plz

**Review by BraulioV — CHANGES_REQUESTED** [review]

**Comment by Rebits** [comment]
### Update 21/05/2021

New changes in Wazuh has produced query tests fails. This may be produced due to a bad predicate values for query in macOS tests, further investigation is required

https://ci.wazuh.info/job/Test_integration/3891/consoleFull

**Comment by Rebits** [comment]
### 24/05/2021

Some minor fixes have been done in query and only-future-events tests. Now, these tests work correctly local environment.
Otherwise, in jenkins, these tests fails in some cases due to a permissions error. These maybe are related with a bad implementation of services tools. Futher investigation is required in order to ensure the correct implementation of these new tests.

**Comment by Rebits** [comment]
### Update 26/05/2021

Most of the tests run correctly in Jenkins environment https://ci.wazuh.info/job/Test_integration/4356/consoleFull 

Some of the tests that fail are:

- **CentOS agent**: `test_basic_configuration_target.py`
- **Windows agent**:
   -  `test_age_datetime_changed`: Due to time sync enabled in AWS instances
   -  Some of the basic configuration tests (`test_configuration`) fails erratically due to a possible bug in a bad validation of the Wazuh Windows agent configuration.
   - In some cases `test_reconnect_time` also fails due to difficulties disabling `Event log `channels.
- **MacOS agent**: `test_configuration_target`

It should be necessary more testing in order to determine if these fails are related to a bug in Wazuh agents or in developed tests.

**Comment by Rebits** [comment]
Checked status of logcollector tests in https://github.com/wazuh/wazuh-qa/issues/1394

File: deps/wazuh_testing/wazuh_testing/logcollector.py [hunk 6]
```diff
@@ -521,8 +562,19 @@ def delete_file_structure(get_files_list):
     Args:
         get_files_list(dict):  Files to delete.
     """
-    for file in get_files_list:
-        shutil.rmtree(file['folder_path'], ignore_errors=True)
+
+    def remove_readonly(func, path, _):
+        "Clear the readonly bit and reattempt the removal"
```
Comment (BraulioV) [inline deps/wazuh_testing/wazuh_testing/logcollector.py:567]:
```suggestion
    def remove_read_only(func, path, _):
        """Clear the read_only bit and reattempt the removal
        
        Args:
        ...
        """
```

File: deps/wazuh_testing/wazuh_testing/logcollector.py [hunk 7]
```diff
@@ -521,8 +562,19 @@ def delete_file_structure(get_files_list):
...
+        os.chmod(path, stat.S_IWRITE)
+        func(path)
+
+    for folder in get_files_list:
+        for i in range(0, 5):
```
Comment (BraulioV) [inline deps/wazuh_testing/wazuh_testing/logcollector.py:572]:
```suggestion
        for _ in range(5):
```

File: deps/wazuh_testing/wazuh_testing/logcollector.py (hunk 6 above)
Comment (Rebits) [reply deps/wazuh_testing/wazuh_testing/logcollector.py]: Done in acb2e52280b139c60e2178171fd356f5100ba6c1

File: deps/wazuh_testing/wazuh_testing/logcollector.py (hunk 7 above)
Comment (Rebits) [reply deps/wazuh_testing/wazuh_testing/logcollector.py]: Done in

**Review by spothound — APPROVED** [review]

**Review by BraulioV — APPROVED** [review]

**Pull request merged**